Lib/test/libregrtest/run_workers.py

```diff
@@ -511,14 +518,18 @@ def _get_result(self) -> QueueOutput | None:
 
         # bpo-46205: check the status of workers every iteration to avoid
         # waiting forever on an empty queue.
-        while any(worker.is_alive() for worker in self.workers):
+        while self.live_worker_count > 0:
```

Comment:
worker.is_alive() is alive is used in other places, I would prefer to use the same logic to list alive workers in all places. I would prefer that you remove this `live_worker_count` attribute.

If any(worker.is_alive() for worker in self.workers) is inefficient, we can design something else, like a list of alive workers and trim this list when a worker exits. But there number of workers should be less than 1000, so it should be fine.

Reply:
I don't understand what design you'd like. We can't continue to rely on the [`is_alive()`](https://docs.python.org/3/library/threading.html#threading.Thread.is_alive) method for the termination condition because that is the source of the race condition: it may happen before or after the processing of the results queue.

There are only two other uses of `is_alive()`, but it makes sense in those places, but doesn't make sense here.

In `__repr__`, where use it for displaying the worker status.

https://github.com/python/cpython/blob/9a1e55b8c5723206116f7016921be3937ef2f4e5/Lib/test/libregrtest/run_workers.py#L118-L119

In `wait_stopped()` it's paired with `join()`. Only `is_alive()` makes sense here because of the use of `join()`.

https://github.com/python/cpython/blob/9a1e55b8c5723206116f7016921be3937ef2f4e5/Lib/test/libregrtest/run_workers.py#L428-L430

Reply:
You can ignore my comment, your change is fine.

I'm thinking at get_running() function, but I was wrong: this function doesn't call the is_alive() method but only relies on `worker.test_name` attribute to decide if a thread is "running" or not. There is already a nested `finally: self.test_name = None` which makes sure that the attribute is cleared in all cases.